torch/fx/immutable_collections.py

```diff
@@ -18,101 +18,105 @@
-        f"'{type(self).__name__}' object does not support mutation. {_help_mutation}",
+
+def _no_mutation(self: Any, *args: Any, **kwargs: Any) -> NoReturn:
+    raise TypeError(
+        f"{type(self).__name__!r} object does not support mutation. {_help_mutation}",
     )
 
 
-def _create_immutable_container(base, mutable_functions):
-    container = type("immutable_" + base.__name__, (base,), {})
-    for attr in mutable_functions:
-        setattr(container, attr, _no_mutation)
-    return container
-
-
-immutable_list = _create_immutable_container(
-    list,
-    (
-        "__delitem__",
-        "__iadd__",
-        "__imul__",
-        "__setitem__",
-        "append",
-        "clear",
-        "extend",
-        "insert",
-        "pop",
-        "remove",
-        "reverse",
-        "sort",
-    ),
-)
-immutable_list.__reduce__ = lambda self: (immutable_list, (tuple(iter(self)),))
-immutable_list.__hash__ = lambda self: hash(tuple(self))
-
-compatibility(is_backward_compatible=True)(immutable_list)
-
-immutable_dict = _create_immutable_container(
-    dict,
-    (
-        "__delitem__",
-        "__ior__",
-        "__setitem__",
-        "clear",
-        "pop",
-        "popitem",
-        "setdefault",
-        "update",
-    ),
-)
-immutable_dict.__reduce__ = lambda self: (immutable_dict, (iter(self.items()),))
-immutable_dict.__hash__ = lambda self: hash(tuple(self.items()))
-compatibility(is_backward_compatible=True)(immutable_dict)
+@compatibility(is_backward_compatible=True)
+class immutable_list(list[_T]):
+    """An immutable version of :class:`list`."""
 
+    __delitem__ = _no_mutation
+    __iadd__ = _no_mutation
+    __imul__ = _no_mutation
+    __setitem__ = _no_mutation
+    append = _no_mutation
+    clear = _no_mutation
+    extend = _no_mutation
+    insert = _no_mutation
+    pop = _no_mutation
+    remove = _no_mutation
+    reverse = _no_mutation
+    sort = _no_mutation
 
-# Register immutable collections for PyTree operations
-def _immutable_dict_flatten(d: dict[Any, Any]) -> tuple[list[Any], Context]:
-    return _dict_flatten(d)
+    def __hash__(self) -> int:  # type: ignore[override]
+        return hash(tuple(self))
 
+    def __reduce__(self) -> tuple[type[Self], tuple[tuple[_T, ...]]]:
+        return (type(self), (tuple(self),))
 
-def _immutable_dict_unflatten(
-    values: Iterable[Any],
-    context: Context,
-) -> dict[Any, Any]:
-    return immutable_dict(_dict_unflatten(values, context))
+
+@compatibility(is_backward_compatible=True)
+class immutable_dict(dict[_KT, _VT]):
+    """An immutable version of :class:`dict`."""
+
+    __delitem__ = _no_mutation
+    __ior__ = _no_mutation
+    __setitem__ = _no_mutation
+    clear = _no_mutation
+    pop = _no_mutation
+    popitem = _no_mutation
+    setdefault = _no_mutation
+    update = _no_mutation  # type: ignore[assignment]
+
+    def __hash__(self) -> int:  # type: ignore[override]
+        return hash(tuple(self.items()))
```

Comment:
`__hash__` and `__eq__` do not match now, because `__hash__` is now order sensitive while `__eq__` is not, right?

Reply:
might need to be a hash of frozenset instead of `self.items()` instead?

Reply:
Updated.